ddl/backfilling.go

```diff
@@ -520,107 +520,144 @@ func splitTableRanges(t table.PhysicalTable, store kv.Storage, startKey, endKey
+	keeper := newDoneTaskKeeper(start)
+	handledTaskCnt := 0
+	for {
+		select {
+		case result := <-scheduler.resultCh:
+			handleOneResult(result, scheduler, consumer, keeper, totalAddedCount, handledTaskCnt)
+			handledTaskCnt++
+		case <-consumer.doneCh:
+			for i := int64(0); i < consumer.taskCnt.Load(); i++ {
+				result := <-scheduler.resultCh
+				handleOneResult(result, scheduler, consumer, keeper, totalAddedCount, handledTaskCnt)
+				handledTaskCnt++
+			}
+			return
+		}
+	}
+}
+
+func handleOneResult(result *backfillResult, scheduler *backfillScheduler, consumer *resultConsumer,
+	keeper *doneTaskKeeper, totalAddedCount *int64, taskSeq int) {
+	reorgInfo := scheduler.reorgInfo
+	if result.err != nil {
+		// Only stores the first error.
+		consumer.err.CompareAndSwap(nil, result.err)
+		logutil.BgLogger().Warn("[ddl] backfill worker failed",
+			zap.Int64("job ID", reorgInfo.ID),
+			zap.String("result next key", hex.EncodeToString(result.nextKey)),
+			zap.Error(result.err))
+		cnt := drainTasks(scheduler.taskCh)
+		consumer.taskCnt.Add(int64(-cnt))
+		return
+	}
+	*totalAddedCount += int64(result.addedCount)
+	keeper.updateNextKey(result.taskID, result.nextKey)
+	if taskSeq%scheduler.workerSize()*4 == 0 {
+		err := consumer.dc.isReorgRunnable(reorgInfo.ID, false)
+		if err != nil {
+			consumer.err.CompareAndSwap(nil, err)
+			logutil.BgLogger().Warn("[ddl] backfill worker is not runnable", zap.Error(err))
+			cnt := drainTasks(scheduler.taskCh)
+			consumer.taskCnt.Add(int64(-cnt))
+			return
+		}
 		failpoint.Inject("MockGetIndexRecordErr", func() {
 			// Make sure this job didn't failed because by the "Write conflict" error.
 			if dbterror.ErrNotOwner.Equal(err) {
 				time.Sleep(50 * time.Millisecond)
 			}
 		})
-		return errors.Trace(err)
+		err = reorgInfo.UpdateReorgMeta(keeper.nextKey, scheduler.sessPool)
+		if err != nil {
+			logutil.BgLogger().Warn("[ddl] update reorg meta failed",
+				zap.Int64("job ID", reorgInfo.ID), zap.Error(err))
+		}
+		// We try to adjust the worker size regularly to reduce
+		// the overhead of loading the DDL related global variables.
+		err = scheduler.adjustWorkerSize()
+		if err != nil {
+			logutil.BgLogger().Warn("[ddl] cannot adjust backfill worker size",
+				zap.Int64("job ID", reorgInfo.ID), zap.Error(err))
+		}
 	}
+}
 
-	metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblOK).Observe(elapsedTime.Seconds())
-	logutil.BgLogger().Info("[ddl] backfill workers successfully processed batch",
-		zap.Stringer("element", reorgInfo.currElement),
-		zap.Int64("total added count", *totalAddedCount),
-		zap.String("start key", hex.EncodeToString(startKey)),
-		zap.String("next key", hex.EncodeToString(nextKey)),
-		zap.Int64("batch added count", taskAddedCount),
-		zap.String("take time", elapsedTime.String()),
-		zap.NamedError("updateHandleError", err1))
-	return nil
+func drainTasks(taskCh chan *reorgBackfillTask) int {
```

Comment:
I suggest that the private function or object can be moved into ```ddl/internal``` and make public. because it is in the ```internal```. It is only public for ```ddl``` package and all subpackage of ```ddl```. It is an effective way to avoid too large ```ddl``` package

Reply:
Removed.